system/settings/storage_bin.c

```diff
@@ -0,0 +1,246 @@
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include "system/settings.h"
+#include <unistd.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <nuttx/crc32.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <nuttx/config.h>
+#include <sys/types.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define VALID          0x600d
+#define BUFFER_SIZE    256     /* Note alignment for Flash writes! */
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+FAR static setting_t *getsetting(char *key);
+extern state_t g_settings;
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: load_bin
+ *
+ * Description:
+ *    Loads binary data from a storage file.
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int load_bin(FAR char *file)
+{
+  static int fd;
+  int i;
+  int ret = OK;
+  uint16_t valid;
+  uint16_t count;
+
+  fd = open(file, O_RDONLY);
+  if (fd < 0)
+    {
+      return -ENOENT;
+    }
+
+  valid = 0;
+  read(fd, &valid, sizeof(uint16_t));
+
+  if (valid != VALID)
+    {
+      goto abort; /* Just exit - the settings aren't valid. What to do? */
+    }
+
+  count = 0;
+  read(fd, &count, sizeof(uint16_t));
+
+  uint32_t calc_crc = 0;
+  uint32_t exp_crc = 0;
+
+  for (i = 0; i < count; i++)
+    {
+      setting_t setting;
+      read(fd, &setting, sizeof(setting_t));
+      calc_crc = crc32part((FAR uint8_t *)&setting, sizeof(setting_t),
+                            calc_crc);
+    }
+
+  read(fd, &exp_crc, sizeof(uint32_t));
+
+  if (calc_crc != exp_crc)
+    {
+      ret = -EIO;
+      goto abort;
+    }
+
+  lseek(fd, (sizeof(uint16_t) *2), SEEK_SET);  /* Get after valid & size */
+
+  for (i = 0; i < count; i++)
+    {
+      setting_t setting;
+      read(fd, &setting, sizeof(setting_t));
+
+      FAR setting_t *slot = getsetting(setting.key);
```

Comment:
move the declaration to the beginning of for loop

Reply:
As above